### ISSUE TITLE
Support for separate debug symbol files when using uprobes?

### ISSUE THREAD

**Issue by jvnn** [iovisor/bpftrace#516]
Just a quick question / feature request: it seems that separate debug files are not supported when using bpftrace with uprobes. The stripped library binary I'm tracing does contain the `.gnu_debuglink` section, and I tried with the debug symbol file in the same directory as the .so file as well as behind Ubuntu's standard /usr/lib/debug/... path, but to no avail. So should this work and I just did something wrong, or is this something that might be supported in the future?

**Comment by brendangregg** [comment]
Some support for debuginfo symbols is supposed to work, but when I tried it didn't work either. Maybe we're missing something. We defer to libbcc for symbol lookups; relevant change:

https://github.com/iovisor/bpftrace/commit/1132d426004faa7fa465ceb4b58d539062afc096

**Comment by chrispsommers** [comment]
**+1 for this feature**. Use-case: want to trace already-running code in a container which is part of a standard "debug" version distro, except the debug files are separate from the main executable. As temporary workaround, I merged the debug symbols back into the executable via elf-utils `eu-unstrip` command, but this was cumbersome in my case. I had to modify the running container, `docker commit` it and restart a bunch of other containers.

BTW I can use `bpftrace -l 'uprobe:path-to-debug-file:symbol'` and find the symbol of interest, but when I try to trace this symbol in a script using same  'uprobe:path-to-debug-file:symbol' it never triggers. I'd expect to just trace `'uprobe:path-to-stripped-binary:symbol'` and let `bpftrace` find the debug file. Does `bpftrace` make some assumptions where the debug file is? In my case its in the ` /usr/lib/debug/.build-id/01/` directory, and `gdb` finds it automatically.

Thanks!

**Comment by danobi** [comment]
> BTW I can use bpftrace -l 'uprobe:path-to-debug-file:symbol' and find the symbol of interest, but when I try to trace this symbol in a script using same 'uprobe:path-to-debug-file:symbol' it never triggers.

Right, this makes sense b/c debug file is never actually executed.

> I'd expect to just trace 'uprobe:path-to-stripped-binary:symbol' and let bpftrace find the debug file.

I agree. This should work.

> Does bpftrace make some assumptions where the debug file is? In my case its in the /usr/lib/debug/.build-id/01/ directory, and gdb finds it automatically.

I haven't looked at the details but this would probably require BCC changes. Shouldn't be too complicated. cc @davemarchevsky

**Comment by chrispsommers** [comment]
Thanks for the responses!

**Comment by deliciouslytyped** [comment]
What is the current status of this?

**Comment by sanmai-NL** [comment]
@chrispsommers 
> **+1 for this feature**. Use-case: want to trace already-running code in a container which is part of a standard "debug" version distro, except the debug files are separate from the main executable. As temporary workaround, I merged the debug symbols back into the executable via elf-utils `eu-unstrip` command, but this was cumbersome in my case. I had to modify the running container, `docker commit` it and restart a bunch of other containers.
> 
> BTW I can use `bpftrace -l 'uprobe:path-to-debug-file:symbol'` and find the symbol of interest, but when I try to trace this symbol in a script using same 'uprobe:path-to-debug-file:symbol' it never triggers. I'd expect to just trace `'uprobe:path-to-stripped-binary:symbol'` and let `bpftrace` find the debug file. Does `bpftrace` make some assumptions where the debug file is? In my case its in the ` /usr/lib/debug/.build-id/01/` directory, and `gdb` finds it automatically.
> 
> Thanks!

Never triggering can also be because the path of the executable somehow doesn't match with how the symbol is invoked. You can test this by artificially invoking the symbol. I couldn't find the exact specs for this executable path argument.

**Comment by jbroadus** [comment]
> **+1 for this feature**. Use-case: want to trace already-running code in a container which is part of a standard "debug" version distro, except the debug files are separate from the main executable. As temporary workaround, I merged the debug symbols back into the executable via elf-utils `eu-unstrip` command, but this was cumbersome in my case. I had to modify the running container, `docker commit` it and restart a bunch of other containers.
> 
> BTW I can use `bpftrace -l 'uprobe:path-to-debug-file:symbol'` and find the symbol of interest, but when I try to trace this symbol in a script using same 'uprobe:path-to-debug-file:symbol' it never triggers. I'd expect to just trace `'uprobe:path-to-stripped-binary:symbol'` and let `bpftrace` find the debug file. Does `bpftrace` make some assumptions where the debug file is? In my case its in the ` /usr/lib/debug/.build-id/01/` directory, and `gdb` finds it automatically.
> 
> Thanks!

This issue is quite old now, but in case anybody stumbled across it with a similar use case --
Since v0.26.0, bcc looks for an env variable BCC_DEBUGINFO_ROOT. If you want to run bpftrace on a process running in a container from outside, the variable can set to the location withing the process's root. For example:
```
export BCC_DEBUGINFO_ROOT="/proc/14110/root/lib/debug"
```
You can then list and use uprobes with the same root path:
```
bpftrace -l "uprobe:/proc/14110/root/bin/dpkg:*"
```

**Comment by jordalgo** [comment]
blazesym, which we are going to use for symbolication, has this ability  - we still need to enable it so I'll leave the issue open for now but blazesym is already integrated in master

**Comment by jschwinger233** [comment]
Can I have try on this? I have always wanted this feature since years ago. :pray:

**Comment by viktormalik** [comment]
@jschwinger233 sure, assigning to you.

Just beware not to do this via BCC as we're slowly trying to get rid of the dependency on BCC.
